src/sites/twitch-twilight/modules/host_button.js

```diff
@@ -0,0 +1,336 @@
+
+		this._chat_con = chatServiceClient;
+		this._chat_con.joinChannel(userLogin);
+	}
+
+	onEnable() {
+		this.metadata.updateMetadata('host');
+
+		const chat = this.resolve('site.chat');
+		chat.ChatController.ready((cls, instances) => {
+			for(const inst of instances) {
+				if (inst && inst.chatService) this.hookIntoChatConnection(inst);
+			}
+		});
+	}
+
+	buildAutoHostMenu(vue, hosts, autoHostSettings, data) {
+		this._current_channel_id = data.id;
+		this.activeTab = this.activeTab || 'auto-host';
+
+		this.vueEl = new vue.Vue({
+			el: e('div'),
+			render: h => h('host-options', {
+				hosts,
+				autoHostSettings,
+				activeTab: this.activeTab,
+
+				addedToHosts: this.currentRoomInHosts(),
+				addToAutoHosts: () => this.addCurrentRoomToHosts(),
+				rearrangeHosts: event => this.rearrangeHosts(event.oldIndex, event.newIndex),
+				removeFromHosts: event => this.removeUserFromHosts(event),
+				setActiveTab: tab => {
+					this.vueEl.$children[0]._data.activeTab = this.activeTab = tab;
+				},
+				updatePopper: () => {
+					if (this._auto_host_tip) this._auto_host_tip.update();
+				},
+				updateCheckbox: e => {
+					const t = e.target,
+						setting = t.dataset.setting;
+					let state = t.checked;
+				
+					if ( setting === 'strategy' )
+						state = state ? 'random' : 'ordered';
+					else if ( setting === 'deprioritize_vodcast' )
+						state = ! state;
+				
+					this.updateAutoHostSetting(setting, state);
+				}
+			})
+		});
+
+		return this.vueEl.$el;
+	}
+
+	async fetchAutoHosts() {
+		const user = this.resolve('site').getUser();
+		if ( ! user )
+			return;
+
+		let data;
+		try {
+			data = await fetch('https://api.twitch.tv/kraken/autohost/list', {
+				headers: {
+					'Accept': 'application/vnd.twitchtv.v4+json',
+					'Authorization': `OAuth ${user.authToken}`
+				}
+			}).then(r => {
+				if ( r.ok )
+					return r.json();
+
+				throw r.status;
+			});
+
+		} catch(err) {
+			this.log.error('Error loading auto host list.', err);
+			return;
+		}
+
+		return this.autoHosts = data.targets;
+	}
+
+	async fetchAutoHostSettings() {
+		const user = this.resolve('site').getUser();
+		if ( ! user )
+			return;
+
+		let data;
+		try {
+			data = await fetch('https://api.twitch.tv/kraken/autohost/settings', {
+				headers: {
+					'Accept': 'application/vnd.twitchtv.v4+json',
+					'Authorization': `OAuth ${user.authToken}`
+				}
+			}).then(r => {
+				if ( r.ok )
+					return r.json();
+
+				throw r.status;
+			});
+
+		} catch(err) {
+			this.log.error('Error loading auto host settings.', err);
+			return;
+		}
+
+		return this.autoHostSettings = data.settings;
+	}
+
+	queueHostUpdate(newHosts) {
+		if (this._host_update_timer) clearTimeout(this._host_update_timer);
+		
+		this._host_update_timer = setTimeout(() => {
+			this._host_update_timer = undefined;
+			this.updateAutoHosts(newHosts);
+		}, 1000);
+	}
+	
+	rearrangeHosts(oldIndex, newIndex) {
+		const newHosts = this.autoHosts.slice(0);
```

Comment:
Is it really necessary to create a new list rather than just splicing an element out of the old list and then adding it back into the old list?